File: cs445/threading/src/gus/synchronize/SafeAccount.java
package gus.synchronize;

public class SafeAccount implements Account { // Why not just extend TcfAccount?
  private Account account;

  public SafeAccount(Account account) { // What is this design pattern called?
    this.account = account;
  }

  public boolean withdraw(long amount) {
    // What does synchronized(account) mean?
    synchronized(account) { return account.withdraw(amount); }
  }

  public long getBalance() {
    synchronized(account) { return account.getBalance(); }
  }

  public static void main (String [] args) {
    Account account = new SafeAccount(new TcfAccount(1000000));
    new MeanMachine("a", account, 125000);
    new MeanMachine("b", account, 250000);
  }
}

File: cs445/threading/ruby/synchronize/safe_account.rb
require 'thread'
require 'synchronize/account'
require 'synchronize/mean_machine'

#
# ruby synchronize/main.rb safe
# ruby1.9 synchronize/main.rb safe
class SafeAccount
  def initialize(account, semaphore)
    @account = account
    @semaphore = semaphore
  end

  def withdraw!(amount)
    @semaphore.synchronize { @account.withdraw!(amount) }
  end

  def balance
    @semaphore.synchronize { @account.balance }
  end
end
